src/clusterfuzz/_internal/bot/tasks/utasks/uworker_io.py

```diff
@@ -0,0 +1,245 @@
+    entity_key = entity_dict['key']
+    serialized_key = base64.b64decode(bytes(entity_key, 'utf-8'))
+    ndb_key = ndb.Key(serialized=serialized_key)
+    entity = get_entity_with_properties(ndb_key, entity_dict['properties'])
+    uworker_input[name] = UworkerEntityWrapper(entity)
+  return uworker_input
+
+
+def serialize_uworker_input(uworker_input):
+  """Serializes and returns |uworker_input| as JSON. Can handle ndb entities."""
+  serializable = {}
+  ndb_entities = {}
+  for key, value in uworker_input.items():
+    if not isinstance(value, ndb.Model):
+      serializable[key] = value
+      continue
+    ndb_entities[key] = make_ndb_entity_input_obj_serializable(value)
+
+  return json.dumps({'serializable': serializable, 'entities': ndb_entities})
+
+
+def serialize_and_upload_uworker_input(uworker_input, job_type,
+                                       uworker_output_upload_url) -> str:
+  """Serializes input for the untrusted portion of a task."""
+  # Add remaining fields to the input.
+  assert 'job_type' not in uworker_input
+  uworker_input['job_type'] = job_type
+  assert 'uworker_output_upload_url' not in uworker_input
+  uworker_input['uworker_output_upload_url'] = uworker_output_upload_url
+
+  uworker_input = serialize_uworker_input(uworker_input)
+  uworker_input_download_url = upload_uworker_input(uworker_input)
+  return uworker_input_download_url
+
+
+def download_and_deserialize_uworker_input(uworker_input_download_url) -> str:
+  """Downloads and deserializes the input to the uworker from the signed
+  download URL."""
+  data = storage.download_signed_url(uworker_input_download_url)
+  return deserialize_uworker_input(data)
+
+
+def serialize_uworker_output(uworker_output):
+  """Serializes uworker's output for deserializing by deserialize_uworker_output
+  and consumption by postprocess_task."""
+  uworker_output = uworker_output.to_dict()
+  entities = {}
+  serializable = {}
+
+  for name, value in uworker_output.items():
+    if not isinstance(value, UworkerEntityWrapper):
+      serializable[name] = value
+      continue
+    entities[name] = {
+        'key': base64.b64encode(value.key.serialized()).decode(),
+        'changed': value._wrapped_changed_attributes,  # pylint: disable=protected-access
+    }
+  return json.dumps({'serializable': serializable, 'entities': entities})
+
+
+def serialize_and_upload_uworker_output(uworker_output, upload_url):
+  """Serializes |uworker_output| and uploads it to |upload_url."""
+  uworker_output = serialize_uworker_output(uworker_output)
+  storage.upload_signed_url(uworker_output, upload_url)
+
+
+def download_and_deserialize_uworker_output(output_url) -> str:
+  """Downloads and deserializes uworker output."""
+  with tempfile.NamedTemporaryFile() as uworker_output_local_path:
+    if not storage.copy_file_from(output_url, uworker_output_local_path.name):
+      logs.log_error('Could not download uworker output from %s' % output_url)
```

Comment:
I think we prefer f-string over `%`?
I am not 100% sure, but it may be ok to have `logs.log_error('Could not download uworker output from %s', output_url)` for logging?